### PR TITLE
Improved interpolate render

### PR DESCRIPTION
**Pull request by Frivas97**
The goal of this PR is to implement a more "stable" render functionality for videos when interpolating between eval camera poses, avoiding fast and sudden movements to improve NeRFs visualization. This PR adds the following arguments to interpolate render video:

- order_poses: The render poses will be ordered by the Euclidian distance between camera images.
- adjust_frame_rate: if True, the video will be rendered at the given frame rate (the duration of the video and the number of interpolated frames between poses will be adjusted accordingly)
- frame_rate: if adjust_frame_rate is True, the output frame rate will be set by this argument.
- seconds_between_poses: if adjust_frame_rate is True, this argument will control the number of interpolated frames between poses so that the camera takes to move between two poses is constant.

Using existing render mode with --interpolation-steps 20 --seconds 20:

https://github.com/nerfstudio-project/nerfstudio/assets/98962882/d10c8c69-d7d8-4254-9b5c-b449cf11c729

Using new render mode with default values:

https://github.com/nerfstudio-project/nerfstudio/assets/98962882/f98d78bd-23dc-475e-939a-16c21a47b208

(Videos are compressed so they can be shown in the PR).

### CODE REVIEW

File: nerfstudio/cameras/camera_utils.py [hunk 1]
```diff
@@ -201,10 +201,40 @@ def get_interpolated_k(k_a, k_b, steps: int = 10) -> TensorType[3, 4]:
...
+        Ks = np.delete(Ks, idx, axis=0)
+
+    return ordered_poses, ordered_ks
+
+
 def get_interpolated_poses_many(
     poses: TensorType["num_poses", 3, 4],
     Ks: TensorType["num_poses", 3, 3],
     steps_per_transition=10,
+    order_poses=False,
```
Comment (tancik) [inline nerfstudio/cameras/camera_utils.py:237]: Add to docstring

Comment (tancik) [reply nerfstudio/cameras/camera_utils.py:237]: Add type annotation

File: nerfstudio/cameras/camera_utils.py [hunk 2]
```diff
@@ -201,10 +201,40 @@ def get_interpolated_k(k_a, k_b, steps: int = 10) -> TensorType[3, 4]:
     return Ks
 
 
+def get_ordered_poses_and_k(poses, Ks):
```
Comment (tancik) [inline nerfstudio/cameras/camera_utils.py:204]: Add type annotations

File: nerfstudio/cameras/camera_utils.py [hunk 3]
```diff
@@ -201,10 +201,40 @@ def get_interpolated_k(k_a, k_b, steps: int = 10) -> TensorType[3, 4]:
...
+    """
+    Returns ordered poses and intrinsics by euclidian distance between poses.
+
+    Args:
+        poses: list of camera poses
+        Ks: list of camera intrinsics
```
Comment (tancik) [inline nerfstudio/cameras/camera_utils.py:210]: Add `Returns` to docstring

File: nerfstudio/scripts/render.py [hunk 4]
```diff
@@ -313,6 +313,14 @@ class RenderTrajectory:
     """How to save output data."""
     interpolation_steps: int = 10
     """Number of interpolation steps between eval dataset cameras."""
+    order_poses: bool = False
+    """Whether to order camera poses by proximity"""
+    adjust_frame_rate: bool = False
+    """Whether to adjust the frame rate of the output video."""
+    frame_rate: int = 25
+    """Frame rate of the output video."""
```
Comment (tancik) [inline nerfstudio/scripts/render.py:321]: Note that this is only for interpolated paths

File: nerfstudio/scripts/render.py [hunk 5]
```diff
@@ -313,6 +313,14 @@ class RenderTrajectory:
     """How to save output data."""
     interpolation_steps: int = 10
     """Number of interpolation steps between eval dataset cameras."""
+    order_poses: bool = False
+    """Whether to order camera poses by proximity"""
+    adjust_frame_rate: bool = False
+    """Whether to adjust the frame rate of the output video."""
+    frame_rate: int = 25
```
Comment (tancik) [inline nerfstudio/scripts/render.py:320]: Make the default 24fps

File: nerfstudio/scripts/render.py [hunk 6]
```diff
@@ -313,6 +313,14 @@ class RenderTrajectory:
     """How to save output data."""
     interpolation_steps: int = 10
     """Number of interpolation steps between eval dataset cameras."""
+    order_poses: bool = False
+    """Whether to order camera poses by proximity"""
+    adjust_frame_rate: bool = False
+    """Whether to adjust the frame rate of the output video."""
```
Comment (tancik) [inline nerfstudio/scripts/render.py:319]: Note that this is only for interpolated paths

**Review by tancik — COMMENTED** [review]
Looks cool! Not needed in this PR, but the more we keep adding functionality to the render script, the more I'm thinking we should split up the commands similar to what we do for `ns-process-data`. ie, `ns-render trajectory ...`

File: nerfstudio/cameras/camera_utils.py (hunk 1 above)
Comment (Frivas97) [reply nerfstudio/cameras/camera_utils.py]: Done

File: nerfstudio/cameras/camera_utils.py (hunk 2 above)
Comment (Frivas97) [reply nerfstudio/cameras/camera_utils.py]: Done

File: nerfstudio/cameras/camera_utils.py (hunk 3 above)
Comment (Frivas97) [reply nerfstudio/cameras/camera_utils.py:213]: Done

File: nerfstudio/scripts/render.py (hunk 5 above)
Comment (Frivas97) [reply nerfstudio/scripts/render.py]: Done

File: nerfstudio/scripts/render.py (hunk 6 above)
Comment (Frivas97) [reply nerfstudio/scripts/render.py]: Done

**Review by tancik — APPROVED** [review]
LGTM

**Pull request merged**